### PR TITLE
Mark local variable as const

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

* Fix [RCS1118: Mark local variable as const](https://github.com/JosefPihrt/Roslynator/blob/master/docs/analyzers/RCS1118.md)
* Inline various constants
* Extract various constants

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs [hunk 1]
```diff
@@ -374,12 +373,6 @@ private TableRowDefinition LoadRowEntryDefinition(XmlNode rowEntryNode, int inde
                     }
                     else if (MatchNodeName(n, XmlTags.TableColumnItemsNode))
                     {
-                        if (columnEntriesNodeFound)
-                        {
-                            this.ProcessDuplicateNode(n);
-                            return null; // fatal
-                        }
-
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs:382]: If compare with multiLineFound the code should follow `columnEntriesNodeFound = true;`. 
If so it is a bug and we should revert the change and fix in new PR. Otherwise columnEntriesNodeFound should be removed at all.

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 2]
```diff
@@ -705,7 +705,7 @@ internal static bool MatchesModulePath(string modulePath, string requiredPath)
 #if UNIX
             StringComparison strcmp = StringComparison.Ordinal;
 #else
-            StringComparison strcmp = StringComparison.OrdinalIgnoreCase;
+            const StringComparison strcmp = StringComparison.OrdinalIgnoreCase;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: Unix case should be fixed too.

File: src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs [hunk 3]
```diff
@@ -1196,10 +1196,10 @@ internal static RemoteDataObject GenerateCreatePowerShell(ClientRemotePowerShell
             // Add Reason property
             if (stateInfo.Reason != null)
             {
-                // If Reason is of not type IContainsErrorRecord, a new ErrorRecord is
-                // created using this errorId
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs:1200]: Why do we remove the comment?

File: src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs [hunk 4]
```diff
@@ -1515,12 +1515,10 @@ internal static Version GetPSRemotingProtocolVersion(RunspacePool rsPool)
             // Add exception property
             if (stateInfo.Reason != null)
             {
-                // If Reason is of not type IContainsErrorRecord,
-                // a new ErrorRecord is created using this errorId
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs:1519]: The same.

File: src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs:382]: The code is unreachable as `columnEntriesNodeFound` is always false, therefore it would seem this is a bug.

I'm not familar with this code but I'm guessing we want something like this in new PR?

```csharp
if (columnEntriesNodeFound)
{
    this.ProcessDuplicateNode(n);
    return null; // fatal
}
columnEntriesNodeFound = true;
```

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 2 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: `StrCmp`? or maybe `IgnoreCaseOpt`?

File: src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs [hunk 5]
```diff
@@ -3571,56 +3566,6 @@ private void SetQuotas()
 
                 Hashtable quotas = transportOption.ConstructQuotasAsHashtable();
 
-                int idleTimeOut = 0;
-
-                if (idleTimeOut != 0 && quotas.ContainsKey(WSManConfigurationOption.AttribMaxIdleTimeout))
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:3576]: @iSazonov Is unreachable code a bug?

File: src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs (hunk 3 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs:1200]: The logic is encapsulated in `GetExceptionProperty` so the comment is moved to method remarks there.

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 2 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: `strCmp` is probably a good option

File: src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs (hunk 1 above)
Comment (rjmholt) [reply src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs:382]: Yes, please revert any bugfixes/semantic changes in this PR. Since this one is quite large, it would be better to keep it focused on the changes described in the title. How to implement bugfixes can be discussed in the new PR

File: src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs (hunk 5 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:3576]: I think so, yes

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: Our code convention is to start with upper case for const.

File: src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:3576]: @xtqqczze Please create new issues for such cases. We could think more and conclude if thus a bug.

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 2 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: My thinking is that that's only for global constants, rather than local constant variables

Comment (iSazonov) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: We should follow CodeFactor and Code Analyzers.

Comment (xtqqczze) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: I am persuaded by @rjmholt's reasoning but our [coding guidelines](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#naming-conventions) state to: "Use `PascalCase to name constant local variables and fields." and is common with [dotnet coding style](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md).

On the other hand, the style of the code in `System.Management.Automation` appears to use mostly `camelCase` over `PascalCase` for local constants and we should generally follow the style of the surrounding code.

`strCmp` would be a violation of [SA1303ConstFieldNamesMustBeginWithUpperCaseLetter](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1303.md) but there is an [open issue](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/2082) as to whether the rule should apply to local constants as well as fields.

I think there could be a seperate discussion on changes to the naming conventions so I am minded not to make this change as part of this PR.

Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: My vote would be for keeping the current variable name unchanged in all cases, and just adding the `const` modifier to the declaration. We can decide later on how local constants should be cased

File: src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs [hunk 6]
```diff
@@ -353,7 +353,7 @@ private TableRowDefinition LoadRowEntryDefinition(XmlNode rowEntryNode, int inde
             using (this.StackFrame(rowEntryNode, index))
             {
                 bool appliesToNodeFound = false;    // cardinality 0..1
-                bool columnEntriesNodeFound = false;         // cardinality 1
+                const bool columnEntriesNodeFound = false;         // cardinality 1
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs:356]: The change causes CI fail because code unreachable after line 379. Possible bug.

File: src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/FormatAndOutput/common/DisplayDatabase/typeDataXmlLoader_Table.cs]: Issue opened: #13246

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 2 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:708]: Issue opened for local constants naming convention: #13247.

File: src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs (hunk 5 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/remoting/commands/CustomShellCommands.cs:3576]: Issue opened: #13248

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 7]
```diff
@@ -260,7 +260,7 @@ private void AppendOneNativeArgument(ExecutionContext context, object obj, Array
         /// <param name="stringConstantType">Bare, SingleQuoted, or DoubleQuoted.</param>
         private void PossiblyGlobArg(string arg, StringConstantType stringConstantType)
         {
-            var argExpanded = false;
+            bool argExpanded = false;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:263]: Unrelated change?

**Review by iSazonov — APPROVED** [review]

**Comment by xtqqczze** [comment]
@iSazonov It seems Roslynator analyzers are not finding diagnostics inside `#if UNIX` etc.

**Review by rjmholt — APPROVED** [review]

**Comment by rjmholt** [comment]
I'll give some extra time for other reviewers. They're likely to be out next week, so we may have to wait a little whlie

**Comment by iSazonov** [comment]
> It seems analyzers do not find code fixes inside an #if directive

@xtqqczze It is expected - the analyzers work at compile. You could target an Unix while run analyzers. But it is not critical.

**Comment by rjmholt** [comment]
@PoshChan please remind me in 10 days

**Comment by iSazonov** [comment]
I checked 99% consts in the repo (local too) is in PascalCase. So we should follow our pattern.

**Comment by xtqqczze** [comment]
rebased to resolve #13316 conflict

**Comment by xtqqczze** [comment]
rebased to resolve merge conflicts

**Comment by daxian-dbw** [comment]
@PowerShell/powershell-maintainers reviewed this PR and agreed to accept it for 7.2 preview. We should delay merging this PR until after **7.1 GA** is done because the chance of confliction caused by this PR is high.

**Comment by iSazonov** [comment]
Set Draft till 7.2 Preview1.

**Comment by iSazonov** [comment]
@rjmholt @SteveL-MSFT I think we can merge.

**Comment by xtqqczze** [comment]
This PR is ready for final review.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Comment by iSazonov** [comment]
@xtqqczze Is it still Draft?

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Review by iSazonov — APPROVED** [review]
